Cython tracebacks are not shown correctly

The following snippet demonstrate a pytest bug in showing Cython tracebacks when we are in
the wrong dir:
https://bitbucket.org/snippets/antocuni/EER6y

How to reproduce::

    $ python setup.py build_ext --inplace
    $ cd testing
    $ py.test --nosugar
    ====================== test session starts ======================
    platform linux2 -- Python 2.7.4 -- py-1.4.30 -- pytest-2.7.1
    rootdir: /data/home/antocuni.large/pypy/misc/pytest-cython-tb, inifile: 
    plugins: sugar
    collected 1 items 

    test_foo.py F

    =========================== FAILURES ============================
    _________________________ test_failing __________________________

        def test_failing():
    >       foo.foo(1, 0)

    test_foo.py:4: 
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
    foo.pyx:1: in foo.foo (foo.c:751)
        ???
    foo.pyx:2: in foo.foo (foo.c:655)
        ???
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    >   ???
    E   ZeroDivisionError: integer division or modulo by zero

    foo.pyx:5: ZeroDivisionError
    =================== 1 failed in 0.01 seconds ====================


Note that the source lines of foo.pyx are not shown.

If we do it outside pytest, it works well::

    $ PYTHONPATH=.. python -c 'import foo; foo.foo(1, 0)'
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "foo.pyx", line 1, in foo.foo (foo.c:751)
        cpdef foo(x, y):
      File "foo.pyx", line 2, in foo.foo (foo.c:655)
        return bar(x, y)+1
      File "foo.pyx", line 5, in foo.bar (foo.c:794)
        return (x*2)/y
    ZeroDivisionError: integer division or modulo by zero

The bug is triggered only when we are inside ``testing``; if we start pytest
from the root dir of the project, the traceback is shown correctly.

--- a/changelog/1139.bugfix.rst
+++ b/changelog/1139.bugfix.rst
@@ -1,0 +1,2 @@
+Fixed tracebacks showing ``???`` instead of source lines for code compiled with a filename relative to a directory other than the current working directory -- e.g. compiled Cython modules, when running pytest from a subdirectory of the project root.
+The source lookup now falls back to searching ``sys.path`` for such filenames, as the standard :mod:`traceback` module does.

--- a/src/_pytest/_code/source.py
+++ b/src/_pytest/_code/source.py
@@ -6,6 +6,7 @@ from bisect import bisect_right
 from collections.abc import Iterable
 from collections.abc import Iterator
 import inspect
+import linecache
 import textwrap
 import tokenize
 import types
@@ -126,7 +127,20 @@ def findsource(obj) -> tuple[Source | None, int]:
     try:
         sourcelines, lineno = inspect.findsource(obj)
     except Exception:
-        return None, -1
+        # inspect.findsource() fails for code objects whose co_filename is
+        # relative to a directory other than the cwd, e.g. compiled Cython
+        # modules when running pytest from a subdirectory (#1139).
+        # Fall back to linecache, which also searches sys.path for bare
+        # filenames, as the standard traceback module does.
+        code = (
+            obj if isinstance(obj, types.CodeType) else getattr(obj, "__code__", None)
+        )
+        if code is None:
+            return None, -1
+        sourcelines = linecache.getlines(code.co_filename)
+        if not sourcelines:
+            return None, -1
+        lineno = code.co_firstlineno - 1
     source = Source()
     source.lines = [line.rstrip() for line in sourcelines]
     source.raw_lines = sourcelines

--- a/testing/code/test_source.py
+++ b/testing/code/test_source.py
@@ -344,6 +344,37 @@ def test_findsource(monkeypatch) -> None:
     assert src[lineno] == "    def x():"
 
 
+def test_findsource_filename_relative_to_syspath_entry(
+    monkeypatch: pytest.MonkeyPatch, tmp_path: Path
+) -> None:
+    """findsource() falls back to searching sys.path for code objects whose
+    co_filename is relative to a directory other than the cwd, like the
+    standard traceback module does (#1139).
+
+    This happens e.g. with compiled Cython modules, whose code objects
+    carry paths relative to the project root, when pytest is run from a
+    subdirectory.
+    """
+    from _pytest._code.source import findsource
+
+    filename = "findsource_syspath_demo.py"
+    lines = ["def f():\n", "    return 1\n"]
+    (tmp_path / filename).write_text("".join(lines), encoding="utf-8")
+    co = compile("".join(lines), filename, "exec")
+    d: dict[str, Any] = {}
+    eval(co, d)
+
+    empty = tmp_path / "empty"
+    empty.mkdir()
+    monkeypatch.chdir(empty)
+    monkeypatch.syspath_prepend(str(tmp_path))
+
+    src, lineno = findsource(d["f"].__code__)
+    assert src is not None
+    assert lineno == 0
+    assert src[lineno] == "def f():"
+
+
 def test_getfslineno() -> None:
     def f(x) -> None:
         raise NotImplementedError()

